larpandoracontent/LArControlFlow/NeutrinoIdTool.cc

```diff
@@ -10,30 +10,464 @@
-            nuSliceHypotheses.at(sliceIndex) : crSliceHypotheses.at(sliceIndex));
+        if (nuSliceHypotheses.at(sliceIndex).size() != 1)
+            throw StatusCodeException(STATUS_CODE_OUT_OF_RANGE);
+
+        const PfoList &nuFinalStates(nuSliceHypotheses.at(sliceIndex).front()->GetDaughterPfoList());
 
-        selectedPfos.insert(selectedPfos.end(), sliceOutput.begin(), sliceOutput.end());
+        CaloHitList reconstructedHits;
+        this->Collect2DHits(nuFinalStates, reconstructedHits);
+        this->Collect2DHits(crSliceHypotheses.at(sliceIndex), reconstructedHits);
+
+        const unsigned int nNuHits(this->CountNeutrinoInducedHits(reconstructedHits));
+        nuNHitsTotal += nNuHits;
+        
+        // TODO may need to be careful with reproducibility here?
+        if (nNuHits > nNuHitsInBestSlice)
+        {
+            nNuHitsInBestSlice = nNuHits;
+            nHitsInBestSlice = reconstructedHits.size();
+            bestSliceIndex = sliceIndex; 
+        }
+    }
+
+    // ATTN for events with no neutrino induced hits, default neutrino purity and completeness to zero
+    float purity(0);
+    float completeness(0);
+
+    if (nHitsInBestSlice > 0)
+        purity = static_cast<float>(nNuHitsInBestSlice) / static_cast<float>(nHitsInBestSlice);
+
+    if (nuNHitsTotal > 0)
+        completeness = static_cast<float>(nNuHitsInBestSlice) / static_cast<float>(nuNHitsTotal);
+
+    return this->PassesQualityCuts(pAlgorithm, purity, completeness);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool NeutrinoIdTool::PassesQualityCuts(const MasterAlgorithm *const pAlgorithm, const float purity, const float completeness) const
+{
+    if (purity < m_minPurity || completeness < m_minCompleteness) return false;
+    if (m_selectNuanceCode && (this->GetNuanceCode(pAlgorithm) != m_nuance)) return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void NeutrinoIdTool::Collect2DHits(const PfoList &pfos, CaloHitList &hitList) const
+{
+    CaloHitList collectedHits;
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_U, collectedHits);
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_V, collectedHits);
+    LArPfoHelper::GetCaloHits(pfos, TPC_VIEW_W, collectedHits);
+
+    for (const CaloHit *const pCaloHit : collectedHits)
+    {
+        // ATTN hits collected from Pfos are copies of hits passed from master instance, we need to access their parent to use MC info
+        const CaloHit *pParentCaloHit(static_cast<const CaloHit *>(pCaloHit->GetParentAddress()));
+
+        // Ensure no hits have been double counted
+        if (std::find(hitList.begin(), hitList.end(), pParentCaloHit) == hitList.end())
```

Comment:
Can find this is incredibly slow, whilst checking a CaloHitSet (filled in parallel) is much faster. Depends on numbers of hits in use, though. We should profile again soon.

Reply:
The number of hits used will range from 10s to 1000s for the MicroBooNE case. Would this simply be a case of replacing CaloHitList -> CaloHitSet?